### PR TITLE
Added e2e upgrade test for Horizontal Pod Autoscaler.

### PR DESCRIPTION
**Pull request by jszczepkowski**
Added e2e upgrade test for Horizontal Pod Autoscaler.

### CODE REVIEW

**Comment by krousey** [comment]
/approve
Aside from build failures, this looks fine to me. I'll let @DirectXMan12 give the final lgtm.

**Comment by jszczepkowski** [comment]
@k8s-bot kops aws e2e test this
issue #37891

**Review by DirectXMan12 — APPROVED** [review]
LGTM :+1:

File: test/e2e/upgrades/horizontal_pod_autoscalers.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+// Teardown cleans up any remaining resources.
+func (t *HPAUpgradeTest) Teardown(f *framework.Framework) {
+	// rely on the namespace deletion to clean up everything
+	t.rc.CleanUp()
+}
+
+func (t *HPAUpgradeTest) test() {
+	t.rc.Resume()
+
+	By(fmt.Sprintf("HPA scales to 1 replica"))
```
Comment (DirectXMan12) [inline test/e2e/upgrades/horizontal_pod_autoscalers.go]: might want a comment specifying how 10 millicores yields 1 replica, 250 yields 3, et

**Review by DirectXMan12 — APPROVED** [review]
small suggested comment, but otherwise LGTM :+1:

File: test/e2e/upgrades/horizontal_pod_autoscalers.go (hunk 1 above)
Comment (jszczepkowski) [inline test/e2e/upgrades/horizontal_pod_autoscalers.go]: done

**Comment by krousey** [comment]
@jszczepkowski Is there a way to speed this up? In my local testing with actual upgrading disabled, most tests complete in under 5 minutes. This one seems to take 25 minutes.

**Comment by skriss** [comment]
@jszczepkowski I'm seeing an error re: this test in some of the CI upgrade jobs:

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gke-container_vm-1.3-container_vm-latest-upgrade-cluster/525?log#log
I0228 21:20:32.860] /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:656
I0228 21:20:32.860]   [k8s.io] cluster upgrade
I0228 21:20:32.860]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:656
I0228 21:20:32.860]     should maintain a functioning cluster [Feature:ClusterUpgrade] [It]
I0228 21:20:32.861]     /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/cluster_upgrade.go:104
I0228 21:20:32.861] 
I0228 21:20:32.861]     Expected error:
I0228 21:20:32.861]         <*errors.StatusError | 0xc4214f0900>: {
I0228 21:20:32.861]             ErrStatus: {
I0228 21:20:32.861]                 TypeMeta: {Kind: "", APIVersion: ""},
I0228 21:20:32.861]                 ListMeta: {SelfLink: "", ResourceVersion: ""},
I0228 21:20:32.861]                 Status: "Failure",
I0228 21:20:32.861]                 Message: "Service \"resource-consumer-upgrade-test\" is invalid: metadata.name: Invalid value: \"resource-consumer-upgrade-test\": must be no more than 24 characters",

**Comment by jszczepkowski** [comment]
@skriss 
Too long name fixed by: https://github.com/kubernetes/kubernetes/pull/42405

**Comment by jszczepkowski** [comment]
@krousey 
> Is there a way to speed this up? In my local testing with actual upgrading disabled, most tests complete in under 5 minutes. This one seems to take 25 minutes.

The long time is caused by HorizontalPodAutoscalerDownscaleForbiddenWindow, which is 5 minutes (HPA can scale down not more than once per 5 minutes). This value is made configurable by https://github.com/kubernetes/kubernetes/pull/42101, so, I after #42101 is merged we should be able to speed it up.

**Pull request merged**